### DISCUSSION TITLE
Use Surge via Cloudflare

### DISCUSSION THREAD

**Discussion by fu-sen** [sintaxi/surge#445] [Ideas]
If you want to use Surge further, Surge Professional is a good choice:
https://surge.sh/pricing

But you might consider using Cloudflare:
- Cloudflare issues an SSL certificate
- Cloudflare CDN exists in an area that is not on the Surge CDN
- Cloudflare also has no limit on the number of site registrations. 
- And Cloudflare offers this for free! (There is also a more ideal paid plan if you use it at work)

Surge can reduce bandwidth. It will give you a little more stable operation, which is also beneficial on the Surge side.

When you go through Cloudflare with each web service, you often run into problems, so we recommend that you do not go through it on the web service side. Surge may also have it. This is a collection of tips for that.

Note: Cloudflare items are many and complex. If you make the wrong settings, your web may not be visible to us at all!

___

**The Cache-Control specified by Surge always references the Surge CDN, even via Cloudflare**:
https://surge.sh/help/using-lucid-caching-automatically
https://support.cloudflare.com/hc/en-us/articles/115003206852-Understanding-Origin-Cache-Control

Still, thanks to the Surge CDN, the response is pleasing in many areas.\
If your region is far from the Surge CDN, you may want to benefit from the Cloudlfare CDN's edge server.\
You can do this with **Page Rules** on the **Cloudflare** side!:

- **example.com/***
- **Edge Cache TTL**
    - **2 hours - a month**

**Cloudflare also doesn't cache `.html` or `sitemap.xml` by default.**\
It's always getting the latest files from the Surge CDN by updating your files.\
That's ideal, but if you still want to cache `.html` and `sitemap.xml`:

- **example.com/***
- **Cache Lebel**
    - **Cache Everything**

Note: After you do this and after updating the file, you will need `Purge Cache` on the Cloudflare side.

___

If you want to apply SSL (TLS) , You can select **Full** for `our SSL/TLS encryption mode`.\
`Full (strict)` cannot be selected, but it can be substituted. (Ideal than Flexible)\
The Surge server usually has a certificate for `* .surge.sh`, Cloudflare can use it.

Are you using Surge Professional? In that case, you can use [Cloudflare Origin CA certificates](https://support.cloudflare.com/hc/en-us/articles/115000479507-Managing-Cloudflare-Origin-CA-certificates).\
[Surge Professional can apply Custom SSL certificate](https://surge.sh/help/securing-your-custom-domain-with-ssl)\
This allows you to select `Full (strict)` in `our SSL / TLS encryption mode`. It's an even better option.

**Comment by sintaxi** [comment]
I discourage the use of CloudFlare with Surge. CloudFlare must check a surge endpoint prior to serving each request which is guaranteed to be slower than using Surge directly. Upgrading to Surge Professional gets you auto-magic cert provisioning/renewal and the CLI has commands for checking and provisioning certs. See #446 for more information about Cert Management on surge.

**Reply by fu-sen** [reply]
Yes, that's the answer you get with other web services as well. Still, there may be reasons why you have to go through Cloudflare. Especially SSL(TLS) application of custom domain. Many web services are now issuing Let's Encrypt, which is probably a priority for Surge. It can eliminate some Cloudflare usage and reduce confusion with Surge support.

**Comment by fu-sen** [comment]
How to make sure that domain goes through Cloudflare
(It can be transferred to Cloudflare support. [Community](https://community.cloudflare.com/) guidance would be ideal)


**Refer to the HTTP header with cURL etc .:**
```
CF-Cache-Status: DYNAMIC
```
You may see `HIT` etc. instead of `DYNAMIC`. and:
```
Server: cloudflare
```
If you can also see `surge-cache:`, it obviously uses Surge.


**See `https://example.com/cdn-cgi/trace`**:
If the domain goes through Cloudflare, it will give a special output:
https://www.cloudflare.com/cdn-cgi/trace


You can also check the WHOIS name server. However, it is not a reliable survey method because you can register for Cloudflare's site on a subdomain if you are using a paid plan. In some cases, Cloudflare is used for DNS-only purposes. (DNS alone is unlimited, so it has an advantage over other name server services)

**Reply by fu-sen** [reply]
I got this with the development version of `surge`. Yes, you are working for that! 💻
```
PS C:\Users\balloon> surge --help

  Surge.sh - Static Web Publishing 0.23.0-rc.4

  surge                             (opt --preview)  publish with prompts
  surge <path>   <domain>           (opt --preview)  publish without prompts (recommended)
  surge encrypt  <domain>                            provision SSL cert for project
  surge certs    <domain>                            view certs for project
  surge config   <domain>                            view/change project configuration
  surge list     <domain>                            list all project revisions
  surge rollfore <domain>                            change to next revision
  surge rollback <domain>                            change to previous revision
  surge cutover  <domain>                            change to latest revision
  surge discard  <revision>                          remove revision from system
  surge files    <domain>                            list all project files
  surge audit    <domain>                            audit edgenode state
  surge bust     <domain>                            busts cache on all edgenodes
  surge traffic  <domain>                            analytics showing project traffic
  surge audience <domain>                            analytics showing audience device info
  surge usage    <domain>                            analytics showing bandwidth usage
  surge teardown <domain>                            tear down a published project
  surge ssl      <domain> <pempath>                  legacy command for uploding .pem file
  surge invite   <domain> <emails>                   invites user to be a contributor
  surge revoke   <domain> <emails>                   revokes contributor rights
  surge dns      <domain>                            view DNS records
  surge dns      <domain> add <type> <name> <value>  add DNS record
  surge dns      <domain> rem <id>                   remove DNS record
  surge whoami                                       show who you are logged in as
  surge login                                        only performs authentication step
  surge logout                                       expire local token
  surge token                                        create token for automation purposes
  surge plan                                         upgrade or downgrade account plan
  surge nuke                                         permanently removes account
  surge list                                         list all projects
  surge --version                                    outputs version
  surge --help                                       outputs this help message

PS C:\Users\balloon> ▄
```

**Reply by fu-sen** [reply]
https://github.com/sintaxi/surge/discussions/446

Can this certificate issuance be targeted at Student (Free Plan)? It depends on whether we choose Surge or another web service. 
Many web services already apply Let ’s Encrypt for Custom domain for free. Even the GitHub Pages that Surge was referring to. It has become common sense. You can give Professional a different degree of freedom to fully differentiate it.

**Reply by fu-sen** [reply]
Surge Professional is used by companies and organizations that have money.These are the ones that require a paid TLS certificate. Individuals who use it as a hobby are usually satisfied with the application of Let’s Encrypt. You can discriminate here by plan.

**Comment by ghost** [comment]
You will get a warning if you dont have SSL. Surge wants me to pay for basic website security.

**Reply by fu-sen** [reply]
This has been Surge's revenue style from the beginning and continues to this day.
This method is still a way for anyone who wants to use Surge.
What the Surge side does not recommend is as stated in the syntaxi comment.

I also have a contract with Netlify and Vercel.
In particular, Vercel has been used since the now.sh era. (It was released at about the same time as Surge)
**Still, there are benefits to using Surge. There is no bandwidth limit and deployment is very quick.**
However, with Surge+Cloudflare, two CDNs overlap, and it can be difficult to support if there is a problem.
Also, free use of Vercel is limited to non-commercial: https://vercel.com/pricing
